instrumentation-api/src/main/java/io/opentelemetry/instrumentation/api/instrumenter/async/AsyncInstrumenter.java

```diff
@@ -0,0 +1,84 @@
+/*
+ * Copyright The OpenTelemetry Authors
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+package io.opentelemetry.instrumentation.api.instrumenter.async;
+
+import io.opentelemetry.context.Context;
+import io.opentelemetry.instrumentation.api.instrumenter.Instrumenter;
+
+/**
+ * A wrapper over {@link Instrumenter} that is able to defer {@link Instrumenter#end(Context,
+ * Object, Object, Throwable)} until asynchronous computation finishes.
+ *
+ * <p>This class is not able to extract {@code RESPONSE} from the asynchronous computation value it
+ * receives in the {@link #asyncEnd(Context, Object, Object, Throwable)} call, so it will always
+ * pass {@code null} as the response to the wrapped {@link Instrumenter}.
+ */
+public final class AsyncInstrumenter<REQUEST> {
+
+  /**
+   * Returns a new {@link AsyncInstrumenter} that wraps over passed {@code syncInstrumenter},
+   * configured for usage with asynchronous computations that are instances of {@code asyncType}.
+   */
+  public static <REQUEST> AsyncInstrumenter<REQUEST> create(
+      Instrumenter<REQUEST, ?> syncInstrumenter, Class<?> asyncType) {
+    return new AsyncInstrumenter<>(
+        syncInstrumenter, asyncType, AsyncEndStrategies.resolveStrategy(asyncType));
+  }
+
+  private final Instrumenter<REQUEST, ?> instrumenter;
+  private final Class<?> asyncType;
+  private final AsyncOperationEndStrategy asyncOperationEndStrategy;
+
+  private AsyncInstrumenter(
+      Instrumenter<REQUEST, ?> instrumenter,
+      Class<?> asyncType,
+      AsyncOperationEndStrategy asyncOperationEndStrategy) {
+    this.instrumenter = instrumenter;
+    this.asyncType = asyncType;
+    this.asyncOperationEndStrategy = asyncOperationEndStrategy;
+  }
+
+  /** Same as calling {@link Instrumenter#shouldStart(Context, Object)}. */
+  public boolean shouldStart(Context parentContext, REQUEST request) {
```

Comment:
Are `shouldStart`/`start` expected to ever be called through `AsyncInstrumenter`?

Reply:
Good question -- it probably won't ever be called in this codebase. To be honest, this class is probably one that I like the least in this PR. Maybe I should have named it `AsyncInstrumenterEnd` and only keep `asyncEnd()` method

Reply:
_Bikeshedding:_  `AsyncOperationEndInstrumenter`, to sync the name with the strategy interface name?

Reply:
👍 
I ended up with `AsyncOperationEndSupport`